File: back/src/main/java/com/back/domain/member/passwordReset/entity/PasswordResetToken.java
package com.back.domain.member.passwordReset.entity;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor
@Table(name = "password_reset_token")
public class PasswordResetToken {

    private static final int MAX_ATTEMPTS = 5;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String email;
    private String code;
    private int failedAttempts;
    private LocalDateTime expiresAt;
    private LocalDateTime createdAt;

    public PasswordResetToken(String email, String code, int expirationMinutes) {
        this.email = email;
        this.code = code;
        this.expiresAt = LocalDateTime.now().plusMinutes(expirationMinutes);
        this.createdAt = LocalDateTime.now();
    }

    public boolean isExpired() {
        return LocalDateTime.now().isAfter(expiresAt);
    }

    public boolean isBlocked() {
        return failedAttempts >= MAX_ATTEMPTS;
    }

    public boolean verifyCode(String inputCode) {
        boolean matched = this.code.equals(inputCode);
        if (!matched) {
            this.failedAttempts++;
        }
        return matched;
    }
}

File: back/src/main/kotlin/com/back/domain/member/passwordReset/entity/PasswordResetToken.kt
package com.back.domain.member.passwordReset.entity

import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.LocalDateTime

@Entity
@Table(name = "password_reset_token")
class PasswordResetToken(
    email: String,
    code: String,
    expirationMinutes: Int
) {
    companion object {
        private const val MAX_ATTEMPTS = 5
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
        protected set

    var email: String = email
        protected set

    var code: String = code
        protected set

    var failedAttempts: Int = 0
        protected set

    var expiresAt: LocalDateTime = LocalDateTime.now().plusMinutes(expirationMinutes.toLong())
        protected set

    var createdAt: LocalDateTime = LocalDateTime.now()
        protected set

    fun isExpired(): Boolean = LocalDateTime.now().isAfter(expiresAt)

    fun isBlocked(): Boolean = failedAttempts >= MAX_ATTEMPTS

    fun verifyCode(inputCode: String): Boolean {
        val matched = code == inputCode
        if (!matched) {
            failedAttempts++
        }
        return matched
    }
}
